### ISSUE TITLE
Proposal: Solution for Deeply Nested Routes

### ISSUE THREAD

**Issue by nhunzaker** [vigetlabs/gangway#23]
## Problem

We need to be able to map deeply nested resources, for example:

```
GET    /members/{id}/notes
GET    /members/{id}/notes/{id}
POST   /members/{id}/notes
PUT    /members/{id}/notes/{:id}
DELETE / members/{id}/notes/{:id}
```

## Proposal

```javascript
API.resource('members', function(members) {
    // Works just like API.resource, only under the 'members' namespace
    members.resource('thing')

    members.resource('notes', function(notes) {
        notes.route({
            aribitraryRoute: { /* options */ }
        })
    })
})
```

### What do you call the ID parameters?

Right now `API.resource` places the ID parameter at `{ params: { id }`. In these cases, we'll have _two_ id values. Should the the first id parameter be `id`, and the second a namespaced key like `resource_id`?

Alternatively, we could do a breaking update to make this more consistent, building paths like:

`/members/{member_id}/notes/{note_id}`

That gets messy because we'd have to get into singularizing... tricky stuff!

**Comment by nhunzaker** [comment]
Something else to think about:

Do we _ever_ use parameters for anything other than IDs? I can't think of a single case where we've done otherwise.

**Comment by ltk** [comment]
Yes, awesome. I prefer the resource-named ids for everything. Inflection is a pretty well-solved problem, just wonder how much weight a good inflection lib would add...

**Comment by ltk** [comment]
only other use for params I can think of is secret tokens for URL-based authentication

**Comment by nhunzaker** [comment]
@ltk aye, I'm less considered with figuring out inflection. Much more worried about the huge tables of english language exceptions :P

**Comment by nhunzaker** [comment]
@ltk Example of secret token url?

**Comment by efatsi** [comment]
I like the way Rails handles nested routes. The deepest object in the nest uses `id` for it's parameter, anything else further up in the chain uses a named `project_id` parameter. Would be a non-breaking change to implement as singular routes follow the same convention.

**Comment by nhunzaker** [comment]
@efatsi I really like that. Now to solve this inflection fun...

**Comment by efatsi** [comment]
https://www.npmjs.com/package/inflected?

**Comment by nhunzaker** [comment]
@efatsi Right. Was just looking at the source actually. Is this list comprehensive enough?

https://github.com/martinandert/inflected/blob/master/lib/defaults.js

Also, cow pluralizes to kline? I had no idea.

**Comment by nhunzaker** [comment]
@efatsi forgive me, the cow/kline thing was in another library:

https://github.com/pksunkara/inflect/blob/master/lib/defaults.js

**Comment by efatsi** [comment]
If it's good enough for Rails, it's good enough for me! https://github.com/rails/rails/blob/master/activesupport/lib/active_support/inflections.rb

**Comment by nhunzaker** [comment]
Yeah. Good enough for me too. Alright cool. Sounds like the next step is simply to build this sucker.

**Comment by nhunzaker** [comment]
Only final sticking point here is that `API.resource` accepts a second argument of options to override the base API config. To avoid a breaking release, we might want to consider doing some type checks at the top of `resource.js`:

https://github.com/vigetlabs/gangway/blob/master/src/resource.js#L5

**Comment by nhunzaker** [comment]
Shipped in 1.3.0
